staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go

```diff
@@ -173,19 +174,40 @@ func validateClaimValidationRules(rules []api.ClaimValidationRule, fldPath *fiel
 	var allErrs field.ErrorList
 
 	seenClaims := sets.NewString()
+	seenExpressions := sets.NewString()
 	for i, rule := range rules {
 		fldPath := fldPath.Index(i)
 
-		if len(rule.Claim) == 0 {
-			allErrs = append(allErrs, field.Required(fldPath.Child("claim"), "claim name is required"))
+		if len(rule.Claim) > 0 {
+			if len(rule.Expression) > 0 {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("expression"), rule.Expression, "claim and expression can't both be set"))
+			}
+			if len(rule.Message) > 0 {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("message"), rule.Message, "message can't be set when claim is set"))
+			}
+			if seenClaims.Has(rule.Claim) {
+				allErrs = append(allErrs, field.Duplicate(fldPath.Child("claim"), rule.Claim))
+			}
+			seenClaims.Insert(rule.Claim)
 			continue
 		}
 
-		if seenClaims.Has(rule.Claim) {
-			allErrs = append(allErrs, field.Duplicate(fldPath.Child("claim"), rule.Claim))
+		if len(rule.Expression) > 0 {
+			// TODO(aramase): do we want to allow empty message?
+			if len(rule.Message) == 0 {
+				allErrs = append(allErrs, field.Required(fldPath.Child("message"), "message is required when expression is set"))
```

Comment:
For api validation, would you consider adding checks for:
1. only spaces in rule.Expression
2. not allow rule.Message contains new line 
3. only spaces in rule.Message
4. Compile if having rule.Expression set, so it could catch compilation err way early in the process.

Reply:
+1 to compiling the message in validation, mentioned at https://github.com/kubernetes/kubernetes/pull/121078#discussion_r1357144745